library/core/src/fmt/num.rs

```diff
@@ -232,83 +233,84 @@ macro_rules! impl_Display {
-                // non-negative, this means that `curr > 0` so `buf_ptr[curr..curr + 1]`
-                // is safe to access.
-                unsafe {
-                    // need at least 16 bits for the 4-characters-at-a-time to work.
-                    #[allow(overflowing_literals)]
-                    #[allow(unused_comparisons)]
-                    // This block will be removed for smaller types at compile time and in the worst
-                    // case, it will prevent to have the `10000` literal to overflow for `i8` and `u8`.
-                    if core::mem::size_of::<$unsigned>() >= 2 {
-                        // eagerly decode 4 characters at a time
-                        while self >= 10000 {
-                            let rem = (self % 10000) as usize;
-                            self /= 10000;
-
-                            let d1 = (rem / 100) << 1;
-                            let d2 = (rem % 100) << 1;
-                            curr -= 4;
-
-                            // We are allowed to copy to `buf_ptr[curr..curr + 3]` here since
-                            // otherwise `curr < 0`. But then `n` was originally at least `10000^10`
-                            // which is `10^40 > 2^128 > n`.
-                            ptr::copy_nonoverlapping(lut_ptr.add(d1 as usize), buf_ptr.add(curr), 2);
-                            ptr::copy_nonoverlapping(lut_ptr.add(d2 as usize), buf_ptr.add(curr + 2), 2);
-                        }
-                    }
-
-                    // if we reach here numbers are <= 9999, so at most 4 chars long
-                    let mut n = self as usize; // possibly reduce 64bit math
+            fn _fmt(self, is_nonnegative: bool, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                const MAX_DEC_N: usize = $unsigned::MAX.ilog(10) as usize + 1;
+                // Buffer decimals for $unsigned with right alignment.
+                let mut buf = [MaybeUninit::<u8>::uninit(); MAX_DEC_N];
+                // Count the number of bytes in buf that are not initialized.
+                let mut offset = buf.len();
+                // Consume the least-significant decimals from a working copy.
+                let mut remain = self;
+
+                // Format per four digits from the lookup table.
+                // Four digits need a 16-bit $unsigned or wider.
+                while const { Self::MAX as u128 > 999 } && remain > 999.try_into().unwrap() {
+                    // SAFETY: All of the decimals fit in buf due to MAX_DEC_N.
+                    unsafe { core::hint::assert_unchecked(offset >= 4) }
+                    // SAFETY: The offset counts down from its initial buf.len()
+                    // without underflow due to the previous assertion.
+                    unsafe { core::hint::assert_unchecked(offset <= buf.len()) }
+                    offset -= 4;
+
+                    let scale: Self = 100_00.try_into().unwrap();
```

Comment:
```suggestion
                    let scale: Self = 10_000.try_into().unwrap();
```

I think the `_` is in the wrong spot

Reply:
The intention was to make the two digit-pairs explicit. Would 1_00_00 be any better?

Reply:
Ah, makes sense. `1_00_00` would probably make it more clear that this is intentional, maybe something like ``// scale shifts `remain` by two two-digit pairs`` to hint why it is like that.

Could you pull `scale` out of the loop too since it is invariant but has some function calls?

Reply:
Yes, will do, including comment.

Scale out of the loop will overflow for i8 & u8. None of these functions are actually called. The compiler makes the number intrinsic. 🙂

Reply:
Yeah, it's trivial for the compiler to licm but pulling it out would still be ideal for tests running in debug mode. Does `$unsigned::try_from(10000u16).unwrap_or($unsigned::MAX)` work?

Reply:
On second thought, 🤷‍♂️ it won't make any noticeable difference